client/filestore-client/src/CordovaFileUpload.ts

```diff
@@ -0,0 +1,62 @@
+import * as Bluebird from 'bluebird';
+
+// Cordova mocks for typescript
+declare var FileUploadOptions: any;
+declare var FileTransfer: any;
+
+// Handling file upload to server
+function fileUpload(fileURI, serverURI, fileUploadOptions) {
+  const transfer = new FileTransfer();
+  return new Bluebird(function(resolve, reject) {
+    transfer.upload(fileURI, serverURI, function(response) {
+      resolve(response);
+    }, function uploadFailure(error) {
+      reject(error);
+    }, fileUploadOptions);
+  });
+}
+
+// Handling retry mechanism of the file upload.
+function fileUploadRetry(fileURI, serverURI, fileUploadOptions, timeout, retries) {
+  return fileUpload(fileURI, serverURI, fileUploadOptions)
+    .then(function(response) {
+      return response;
+    }, function() {
+      if (retries === 0) {
+        throw new Error('Cannot upload to file');
+      }
+      return Bluebird.delay(timeout)
+        .then(function() {
+          return fileUploadRetry(fileURI, serverURI, fileUploadOptions, timeout, retries - 1);
+        });
+    });
+}
+
+/**
+ * Upload file using local file URI. Used for uploads on mobile devices (cordova based)
+ *
+ * @param baseUrl
+ * @param userId
+ * @param fileURI
+ * @param options
+ * @returns {*}
+ */
+export function uploadFile(baseUrl, userId, fileURI, options) {
+  if (arguments.length < 2) {
+    return Bluebird.reject('userId and fileURI parameters are required.');
+  } else {
+    options = options || {};
+    const fileUploadOptions = new FileUploadOptions();
+    fileUploadOptions.fileKey = options.fileKey || 'binaryfile';
+    fileUploadOptions.fileName = options.fileName;
+    fileUploadOptions.mimeType = options.mimeType || 'image/jpeg';
+    fileUploadOptions.params = {
+      ownerId: userId,
+      fileName: options.fileName
+    };
+    const timeout = options.timeout || 2000;
+    const retries = options.retries || 1;
+    const serverURI = baseUrl + '/upload/binary';
```

Comment:
This is something I will need to extract and improve now. 
Actual network client needs to come from developers as they need to apply some headers for security etc.

@paolobueno - Do you think we can put that into angularjs-http module? Thinking how to get that without creating additional artifacts.

Reply:
Makes sense to reutilize our own angularjs-http module. I think this module will need to receive a 'compatible' http client as a dependency and defer to it for the actual communications.

I've taken a look at the file transfer plugin and I think it's not required for this, and we can use any http lib instead as long as we can configure the correct mime types and headers: https://github.com/apache/cordova-plugin-file-transfer#deprecated

We probably need to add support for file uploads to the angularjs-http module, since it was built only as a json-based REST client. This can be done like this: https://stackoverflow.com/a/25264008/2600345, maybe adding something like `HttpDataService.prototype.uploadFile(relativeUrl, fileData)`

What I'm worried about right now is introducing a circular dependency between both core and angularjs repositories.

Reply:
A way to avoid coupling this module with `angularjs-http` would be for it to only receive a simple callback `uploadFile(url, data) => Promise` and let the demo application/step configure it to come from the angularjs module.

Reply:
> What I'm worried about right now is introducing a circular dependency between both core and angularjs repositories.

Let's bind them in the reference app so it will not be hard dependency.